modules/curl/curl.c

```diff
@@ -0,0 +1,311 @@
+
+  if (self->body_template) {
+    body_rendered = g_string_new(NULL);
+    log_template_format(self->body_template, msg, &self->template_options, LTZ_SEND,
+                        self->super.seq_num, NULL, body_rendered);
+    body = body_rendered->str;
+  } else
+    body = log_msg_get_value(msg, LM_V_MESSAGE, NULL);
+
+  curl_easy_reset(self->curl);
+
+  curl_easy_setopt(self->curl, CURLOPT_WRITEFUNCTION, _curl_write_cb);
+
+  curl_easy_setopt(self->curl, CURLOPT_URL, self->url);
+
+  if (self->user)
+    curl_easy_setopt(self->curl, CURLOPT_USERNAME, self->user);
+
+  if (self->password)
+    curl_easy_setopt(self->curl, CURLOPT_PASSWORD, self->password);
+
+  curl_easy_setopt(self->curl, CURLOPT_USERAGENT, self->user_agent);
+
+  g_snprintf(header_host, sizeof(header_host),
+    "X-Syslog-Host: %s", log_host);
+  curl_headers = curl_slist_append(curl_headers, header_host);
+
+  g_snprintf(header_program, sizeof(header_program),
+    "X-Syslog-Program: %s", log_program);
+  curl_headers = curl_slist_append(curl_headers, header_program);
+
+  g_snprintf(header_facility, sizeof(header_facility),
+    "X-Syslog-Facility: %s", log_facility);
+  curl_headers = curl_slist_append(curl_headers, header_facility);
+
+  g_snprintf(header_level, sizeof(header_level),
+    "X-Syslog-Level: %s", log_level);
+  curl_headers = curl_slist_append(curl_headers, header_level);
+
+  header = self->headers;
+  while (header != NULL) {
+    curl_headers = curl_slist_append(curl_headers, (gchar *)header->data);
+    header = g_list_next(header);
+  }
+
+  curl_easy_setopt(self->curl, CURLOPT_HTTPHEADER, curl_headers);
+
+  curl_easy_setopt(self->curl, CURLOPT_POSTFIELDS, body);
+
+  if ((ret = curl_easy_perform(self->curl)) != CURLE_OK) {
+      msg_error("curl: error sending HTTP request",
+                evt_tag_str("error", curl_easy_strerror(ret)),
+                NULL);
+
+      if (body_rendered)
+        g_string_free(body_rendered, TRUE);
+
+      curl_slist_free_all(curl_headers);
+
+      return WORKER_INSERT_RESULT_ERROR;
+  }
+
+  if (body_rendered)
+    g_string_free(body_rendered, TRUE);
+
+  curl_slist_free_all(curl_headers);
+
+  return WORKER_INSERT_RESULT_SUCCESS;
+}
+
+void
+curl_dd_set_url(LogDriver *d, const gchar *url)
+{
+  CurlDestinationDriver *self = (CurlDestinationDriver *) d;
+
+  g_free(self->url);
+  self->url = g_strdup(url);
+}
+
+void
+curl_dd_set_user(LogDriver *d, const gchar *user)
+{
+  CurlDestinationDriver *self = (CurlDestinationDriver *) d;
+
+  g_free(self->user);
+  self->user = g_strdup(user);
+}
+
+void
+curl_dd_set_password(LogDriver *d, const gchar *password)
+{
+  CurlDestinationDriver *self = (CurlDestinationDriver *) d;
+
+  g_free(self->password);
+  self->password = g_strdup(password);
+}
+
+void
+curl_dd_set_user_agent(LogDriver *d, const gchar *user_agent)
+{
+  CurlDestinationDriver *self = (CurlDestinationDriver *) d;
+
+  g_free(self->password);
+  g_stpcpy(self->user_agent, user_agent);
```

Comment:
You should check for the size of the input to avoid writing beyond the self->user_agent `char[1024]` buffer limit – use `g_strlcpy()` instead?